xtokens/src/lib.rs

```diff
@@ -467,43 +469,140 @@ pub mod module {
+			// In case that fee reserve != asset reserve, there are two xcm sent from
+			// sender. first xcm sent to fee reserve chain which route to dest chain. second
+			// xcm directly sent to dest chain. the fee amount in fee asset is split into
+			// two parts. 1. fee asset sent to fee reserve chain = fee_amount - min_xcm_fee
+			// 2. fee asset sent to dest reserve chain = min_xcm_fee
+			// the first part + second part = fee amount in fee asset
+			let fee_reserve = fee.reserve();
+			if fee_reserve != non_fee_reserve {
+				// Current only support `ToReserve` with relay-chain asset as fee. other case
+				// like `NonReserve` or `SelfReserve` with relay-chain fee is not support.
+				ensure!(non_fee_reserve == dest.chain_part(), Error::<T>::InvalidAsset);
+
+				let reserve_location = non_fee_reserve.clone().ok_or(Error::<T>::AssetHasNoReserve)?;
+				let min_xcm_fee = T::MinXcmFee::get(&reserve_location);
+
+				// min xcm fee should less than user fee
+				let fee_to_dest = reset_fee(&fee, min_xcm_fee);
+				ensure!(fee_to_dest < fee, Error::<T>::InvalidAsset);
+
+				let mut assets_to_dest = MultiAssets::new();
+				for i in 0..asset_len {
+					let asset = assets.get(i).ok_or(Error::<T>::AssetIndexNonExistent)?;
+					if fee != *asset {
+						assets_to_dest.push(asset.clone());
+					} else {
+						assets_to_dest.push(fee_to_dest.clone());
+					}
 				}
-				ToNonReserve => Self::transfer_to_non_reserve(
+
+				let mut assets_to_fee_reserve = MultiAssets::new();
+				let asset_to_fee_reserve = subtract_fee(&fee, min_xcm_fee);
+				assets_to_fee_reserve.push(asset_to_fee_reserve.clone());
+
+				// First xcm sent to fee reserve chain and routed to dest chain.
+				// The `SelfLocation` current is (1, Parachain(id)) refer to sender parachain.
+				// we use `SelfLocation` to fund fee to sender's parachain sovereign account on
+				// destination chain, which asset is originated from sender account on sender
+				// chain. This means if user setup too much fee, the fee is not returned to
+				// user, instead deposit to sibling parachain sovereign account on dest chain.
+				// Notice: if parachain set `SelfLocation` to (0, Here), it'll be error!
+				Self::send_xcm(
+					origin_location.clone(),
+					assets_to_fee_reserve,
+					asset_to_fee_reserve,
+					fee_reserve,
+					&dest,
+					Some(T::SelfLocation::get()),
+					dest_weight,
+				)?;
+
+				// Second xcm send to dest chain.
+				// Current not ensure xcm order delivery. if second xcm is executed before first
+				// xcm, then second xcm may failed because of sibling parachain account don't
+				// have enough fee to withdraw. we can pre-fund some amount to sibling parachain
+				// sovereign account to fix this issue. when first xcm executed later on, the
+				// sibling sovereign parachain account is deposit. and next transaction will
+				// succeed even though second xcm is executed before first xcm if user fee is
+				// less than parachain sovereign account balance.
```

Comment:
Dispatchable calls docstring or pallet readme should include these usage notes.

Reply:
move to README